### DISCUSSION TITLE
publicClient.simulateContract({...}) throw the Error with opcode -32000

### DISCUSSION THREAD

**Discussion by DmitriiKholodov** [wevm/viem#833] [Question] [answered]
Hello! I have an issue with preparing contract.

I am trying to simulateContract before I will prepare and write it.
I have a simple Wagmi config which I pass to <WagmiProvider />

```typescript
const { chains, publicClient, webSocketPublicClient } = configureChains(
  [mainnet],
  [alchemyProvider({ apiKey: process.env.REACT_APP_ALCHEMY_PUBLIC_KEY ?? '' }), publicProvider()],
);

export const config = createConfig({
  autoConnect: true,
  connectors: [
    new MetaMaskConnector({ chains }),
    new WalletConnectConnector({
      chains,
      options: {
        projectId: process.env.REACT_APP_WALLET_CONNECT_PROJECT_ID ?? '',
      },
    }),
  ],
  publicClient,
  webSocketPublicClient,
});
``` 

Here's I am trying to simulateContract

const publicClient = usePublicClient();

```typescript
publicClient.simulateContract({
    address: '0xdAC17F958D2ee523a2206206994597C13D831ec7', // USDT token address
    abi: erc20ABI, // imported from wagmi
    functionName: 'transferFrom',
    args: [
      '0xcFF659874f31Df59835B661b1d7e590E50200889',
      '0xCBB095d832D23b9930EDA3be9329efB91555097E',
      parseEther('0.1'),
    ],
  })
    .then(console.log)
    .catch(console.warn);
```

So I get the error

```typescript
ContractFunctionExecutionError: An error occurred while executing: RPC Request failed.

URL: https://eth-mainnet.g.alchemy.com/v2/...
Request body: {"method":"eth_call","params":[{"data":"0x23b872dd000000000000000000000000cff659874f31df59835b661b1d7e590e50200889000000000000000000000000cbb095d832d23b9930eda3be9329efb91555097e000000000000000000000000000000000000000000000000016345785d8a0000","to":"0xdAC17F958D2ee523a2206206994597C13D831ec7"},"latest"]}

Details: invalid opcode: INVALID
Version: viem@1.2.2

Raw Call Arguments:
  to:    0xdAC17F958D2ee523a2206206994597C13D831ec7
  data:  0x23b872dd000000000000000000000000cff659874f31df59835b661b1d7e590e50200889000000000000000000000000cbb095d832d23b9930eda3be9329efb91555097e000000000000000000000000000000000000000000000000016345785d8a0000
 
Contract Call:
  address:   0xdAC17F958D2ee523a2206206994597C13D831ec7
  function:  transferFrom(address sender, address recipient, uint256 amount)
  args:                  (0xcFF659874f31Df59835B661b1d7e590E50200889, 0xCBB095d832D23b9930EDA3be9329efB91555097E, 100000000000000000)

Docs: https://viem.sh/docs/contract/simulateContract.html
Details: invalid opcode: INVALID
Version: viem@1.2.2
    at getContractError (getContractError.js:33:1)
    at simulateContract (simulateContract.js:78:1)
``` 

How can I get more details about the error. It does not looks usefull 

```typescript
{
    "jsonrpc": "2.0",
    "id": 0,
    "error": {
        "code": -32000,
        "message": "invalid opcode: INVALID"
    }
}
``` 

I've tried to google opcode, but I did not find any details.

Could you suggest some actions to solve my issue please?

**Comment by jxom** [comment]
Your Alchemy url doesn’t have an API key. Is it meant to?

**Reply by DmitriiKholodov** [reply]
It has, I just remove that in example

**Comment by MidnightLightning** [comment, marked as answer]
I'm guessing the error here could be that because you're using the viem `publicClient` to do the simulation, it's using some random address as the "from" address rather than the user's connected Metamask/WalletConnect wallet. `invalid opcode: INVALID` is the output I see when I try simulating that same transaction in tenderly (sending from the zero-address):

<img width="1542" alt="Screenshot 2023-07-03 at 8 15 59 AM" src="https://github.com/wagmi-dev/viem/assets/250250/a1bebbf3-cf6a-4d34-ab53-423948bf0018">

Which is the correct error to receive for an ERC20 token, since the zero-address does not have permission to spend the `0xcFF659874f31Df59835B661b1d7e590E50200889` address' tokens.

**Reply by DmitriiKholodov** [reply]
It is exactly what I needed! Thanks a lot!

I guessed that I should use walletClient, but did not know which way. Every example I have seen was with useContractWrite, but as I see now, i should use walletClient.contractWrite

Now It works this way, It triggered my connected MetaMask extension

<img width="767" alt="image" src="https://github.com/wagmi-dev/viem/assets/28441549/9f0049ed-227b-4895-8ce3-48e38017f238">

**Reply by janndriessen** [reply]
Thanks, this thread helped me as well! Just leaving here that this could also be solved adding `account` to `simulateContract`. So you could still use the `publicClient`. So for example:

```typescript
publicClient.simulateContract({
    account: '0x...' // add account address
    address: '0xdAC17F958D2ee523a2206206994597C13D831ec7', // USDT token address
    abi: erc20ABI, // imported from wagmi
    functionName: 'transferFrom',
    args: [
      '0xcFF659874f31Df59835B661b1d7e590E50200889',
      '0xCBB095d832D23b9930EDA3be9329efB91555097E',
      parseEther('0.1'),
    ],
  })
```

This might be something for the docs too!
